core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java

```diff
@@ -0,0 +1,391 @@
+    }
+  }
+
+  private void initializeCatalogTables() throws InterruptedException, SQLException {
+    // need to check multiple times because some databases are using different naming standard. ex: H2db keeping
+    // table names as uppercase
+    boolean exists = false;
+    DatabaseMetaData dbMeta = dbConnPool.run(Connection::getMetaData);
+    ResultSet tables = dbMeta.getTables(null, null, SQL_TABLE_NAME, null);
+    if (tables.next()) {
+      exists = true;
+    }
+    tables.close();
+    ResultSet tablesUpper = dbMeta.getTables(null, null, SQL_TABLE_NAME.toUpperCase(), null);
+    if (tablesUpper.next()) {
+      exists = true;
+    }
+    tablesUpper.close();
+    ResultSet tablesLower = dbMeta.getTables(null, null, SQL_TABLE_NAME.toLowerCase(), null);
+    if (tablesLower.next()) {
+      exists = true;
+    }
+    tablesLower.close();
+
+    // create table if not exits
+    if (!exists) {
+      dbConnPool.run(conn -> conn.prepareStatement(SQL_TABLE_DDL).execute());
+      LOG.debug("Created table {} to store iceberg tables!", SQL_TABLE_NAME);
+    }
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new JdbcTableOperations(dbConnPool, fileIO, catalogName, tableIdentifier);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier table) {
+    if (table.hasNamespace()) {
+      return SLASH.join(warehouseLocation, table.namespace().levels(), table.name());
+    }
+    return SLASH.join(warehouseLocation, table.name());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    TableOperations ops = newTableOps(identifier);
+    TableMetadata lastMetadata = ops.current();
+    try {
+      PreparedStatement sql = dbConnPool.run(c -> c.prepareStatement(SQL_DELETE_TABLE));
+      sql.setString(1, catalogName);
+      sql.setString(2, JdbcUtil.namespaceToString(identifier.namespace()));
+      sql.setString(3, identifier.name());
+      int deletedRecords = sql.executeUpdate();
+
+      if (deletedRecords > 0) {
+        LOG.debug("Successfully dropped table {}.", identifier);
+      } else {
+        throw new NoSuchTableException("Cannot drop table %s! table not found in the catalog.", identifier);
+      }
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+        LOG.info("Table {} data purged!", identifier);
+      }
+      return true;
+    } catch (SQLException | InterruptedException e) {
+      throw new UncheckedSQLException("Failed to drop table!", e);
+    }
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    if (!this.namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace %s does not exist!", namespace);
+    }
+    try {
+      List<TableIdentifier> results = Lists.newArrayList();
+      PreparedStatement sql = dbConnPool.run(c -> c.prepareStatement(SQL_SELECT_ALL));
```

Comment:
I think that the result set and prepared statement are tied to the connection that was used to create the statement. That means that as long as this method is using either the statement or the result set, the connection should not be reused by another thread.

To hold the connection, the whole query will need to go in the `run` block. Here's my version of this method with that change:

```java
  @Override
  public List<TableIdentifier> listTables(Namespace namespace) {
    if (!this.namespaceExists(namespace)) {
      throw new NoSuchNamespaceException("Namespace %s does not exist!", namespace);
    }

    try {
      return dbConnPool.run(conn -> {
        List<TableIdentifier> results = Lists.newArrayList();
        PreparedStatement sql = conn.prepareStatement(SQL_SELECT_ALL);

        sql.setString(1, catalogName);
        sql.setString(2, JdbcUtil.namespaceToString(namespace));

        try (ResultSet rs = sql.executeQuery()) {
          while (rs.next()) {
            results.add(JdbcUtil.stringToTableIdentifier(rs.getString("table_namespace"), rs.getString("table_name")));
          }
        }

        return results;
      });

    } catch (SQLException e) {
      throw new UncheckedSQLException(String.format("Failed to list tables in namespace: %s", namespace), e);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new RuntimeException("Interrupted during JDBC operation", e);
    }
  }
```

I've also separated out the `InterruptedException` and handled it just like in `HiveCatalog`. I think that should be done in all cases here as well.

Reply:
thank you for the example and tip applied it to all cases